### PR TITLE
Generate slugs updated

### PR DESCRIPTION
**Pull request by felipesere**
Initial PR to begin talking about this

Todos:
* [x] Show an error in the form when the slug is already taken

### CODE REVIEW

File: lib/lighthouse/slugger.ex [hunk 1]
```diff
@@ -0,0 +1,8 @@
+defmodule Lighthouse.Slugger do
+  def generate(title) do
+    title
+    |> String.downcase
+    |> String.replace(" ","-")
+    |> String.replace(~r/[^\w-]/,"")
```
Comment (marosluuce) [inline lib/lighthouse/slugger.ex]: Extracting these as methods may make it clearer what this is doing.

File: test/books/controller_test.exs [hunk 2]
```diff
@@ -56,21 +57,29 @@ defmodule Lighthouse.Books.ControllerTest do
       "description" => "d",
       "link"        => "e"
     }
-    post conn(), "/books", %{book: params, category: category.name}
+    post conn(), "/books", %{"book" => params, "category" => category.name}
```
Comment (marosluuce) [inline test/books/controller_test.exs]: Why change this?

File: test/categories/category_for_paper_test.exs [hunk 3]
```diff
@@ -2,9 +2,10 @@ defmodule Lighthouse.Categories.CategoryForPaperTest do
   use Lighthouse.RepositoryCase
   alias Lighthouse.Categories.Category
   alias Lighthouse.Categories.CategoryFor
+  alias Lighthouse.Papers.Query
 
   test "saves a relation to a paper" do
-    paper = Lighthouse.SampleData.sample_paper |> Repo.insert!
+    {:ok, paper} = Lighthouse.SampleData.sample_paper |> Query.save
```
Comment (marosluuce) [inline test/categories/category_for_paper_test.exs]: `sample_paper()`

File: test/books/controller_test.exs (hunk 2 above)
Comment (felipesere) [inline test/books/controller_test.exs]: I had issues with the controller not picking up. I will change it back and see if I got resolved with a different change. Good catch :+1:

**Comment by felipesere** [comment]
This is ready to be merged. Please have a look @christophgockel @Maikon @marosluuce @heruku

File: lib/faros/books/query.ex [hunk 4]
```diff
@@ -5,6 +5,12 @@ defmodule Faros.Books.Query do
 
   def all, do: Repo.all(from b in Book, select: b)
 
+  def save(book_data) do
+    %Book{}
+    |> Book.changeset(book_data)
+    |> Repo.insert
+  end
+
```
Comment (Maikon) [inline lib/faros/books/query.ex]: It's interesting that the Query is responsible for creating new records. What's the rationale behind this?

Comment (felipesere) [inline lib/faros/books/query.ex]: Nathan started splitting out queries into `Query` and modifications which went straight to `Faros.Repo`.
By adding the constraint I had to control how to `save` things, and that would have spread.
I want to rename `Query` to something else, maybe `DB`? I could do it as part of this PR.

File: test/books/query_test.exs [hunk 5]
```diff
@@ -4,20 +4,25 @@ defmodule Faros.Books.QueryTest do
...
+    {:ok, book} = sample_book() |> Query.save
 
-    assert saved_book.slug == book.slug
+    assert sample_book().slug == book.slug
+  end
+
+  test "slugs of books must be unique" do
+    {:ok, _} = sample_book() |> Query.save
+    {:error, changeset} = sample_book() |> Query.save
+    assert changeset.errors[:slug]
```
Comment (Maikon) [inline test/books/query_test.exs]: :+1:

File: test/categories/category_for_book_test.exs [hunk 6]
```diff
@@ -2,9 +2,11 @@ defmodule Faros.Categories.CategoryForBookTest do
   use Faros.RepositoryCase
   alias Faros.Categories.Category
   alias Faros.Categories.CategoryFor
+  alias Faros.Books.Query
+  alias Faros.SampleData
 
   test "saves a relation to a book" do
-    book = Faros.SampleData.sample_book  |> Repo.insert!
+    {:ok, book} = SampleData.sample_book()  |> Query.save
```
Comment (Maikon) [inline test/categories/category_for_book_test.exs]: Maybe it's best importing the `sample_book()` as we do elsewhere so we don't need to prefix with the module? More for consistency.

Comment (felipesere) [inline test/categories/category_for_book_test.exs]: I don't think we do elsewhere but I like the idea a lot! Will do it and push up in a few

File: test/papers/query_test.exs [hunk 7]
```diff
@@ -1,28 +1,34 @@
 defmodule Faros.Papers.QueryTest do
-  import Faros.SampleData, only: [sample_paper: 0]
+  import Faros.SampleData, only: [sample_paper: 0, sample_paper: 1]
```
Comment (Maikon) [inline test/papers/query_test.exs]: Here :)

**Pull request merged**